### ISSUE TITLE
Sentinel-DNS: local threat detection, QoE, and Wazuh over DNS telemetry (Epic)

### ISSUE THREAD

**Issue by KerubeDev** [KerubeDev/Track4#2]
# Sentinel-DNS â€” Build Context (Epic)

> Epic Â· work_type=Epic Â· Milestone: Track4 - Delivery Â· Deliverables: repo + video â‰¤ 5 min in Spanish before Sep 11 2026 08:00 (UTCâˆ’5)

## Goals

Build **Sentinel-DNS**, the solution for the Ovnicom corporate challenge at the Decentralized AI Hackathon (Tether, Sep 2026, Panama City): real-time, explainable intelligence over a corporate DNS telemetry stream â€” entirely on-device via **QVAC**, with structured alerts into **Wazuh** and an interpretable **per-site QoE score**.

The pipeline must deliver, in real time over the query stream:

1. **Threat detection** â€” DGA, DNS tunneling, typosquatting, periodic beaconing â€” explainable, not a black box.
2. **QoE per site** â€” a numeric interpretable score per site / zone / PoP to spot degraded resolution before users complain.
3. **Wazuh integration** â€” every finding lands in Wazuh as a structured, processable alert for the SOC.
4. **Fully local AI** â€” inference runs on-device via **QVAC** (QWEN3 1.7B, precached, port 11434), never on a cloud API. Cloud inference = instant disqualification.

## Non-goals

- No ML model training or fine-tuning; threat analysis uses deterministic signals + local LLM verdicts.
- No changes to DNS resolution itself; the pipeline is read-only over telemetry.
- No cloud-hosted inference or storage of sensitive data.
- The raw synthetic dataset (~721 MB, 36 files) is not committed; a download script plus sample covers reproducibility.

## Architecture (confirmed decisions)

```
BIND dataset â”€â”€â–º emulator (replay P5, attack episodes P4) â”€â”€ dns.telemetry.v1 (Kafka)
        ClickHouse (dns_events_raw + site_qoe_minute) â”€â”€â–º Grafana dashboard (5 panels)
                  â–¼
             agent/filter (5 deterministic signals P1+P1b)
         no signal            any escalation rule met
                  â”‚              â”‚  â”€â”€â–º QVAC (local, verdict JSON)
        benignâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â–º alert: Wazuh via syslog (rules 100101â€“100106)
```

Key confirmed decisions:

| # | Decision |
|---|---|
| P1 | Workflow = 5 deterministic signals (NXDOMAIN ratio, entropy, composite length, rarity, beaconing) over a sliding window per `client_ip` / e2ld, configurable thresholds; only escalation-triggering queries reach QVAC |
| P1b | Signal 3 composite: single label > 40 chars **AND** entropy > 3.5 **AND** repetition â‰¤ 60 s (avoids CDN/UUID false positives) |
| P1-escalation | Signals 1, 3, 5 escalate alone; signal 2 only with 4 (rarity); signal 4 never alone; typosquatting via edit-distance against the client's vocabulary |
| P2 | QVAC verdict contract: strict one-shot JSON `{verdict, confidence, reasoning_short, recommended_action}`; malformed JSON â†’ `unverified` alert |
| P2b | QoE weights 45/35/20 (latency/NXDOMAIN/saturation), labels 85/70/50, justification-guarded in `config/qoe.yaml` |
| P3 | Evaluation: unit = query; TP/FP/FN/TN; precision/recall/F1 (macro + per class); filter elimination rate; p50/p95 latency; split by zone_id/pop_id; ground_truth only in the emulator |
| P3b | Verdictâ†’severity Wazuh mapping in `local_rules.xml`: dgaâ‰¥0.90â†’12, dga<0.90â†’9, tunnelâ†’12, beaconingâ†’10, typosquatâ†’6, unverifiedâ†’5; benign emits no rule |
| P4 | Attack script: 5 escalating episodes (E1 DGA ~2k/5â€“8 IPs, E2 typosquat ~300, E3 tunnel ~1k/1 IP, E4+E5 beaconing ~60 s) over real-dataset background, fixed seed |
| P5 | Dual clock: emulator writes original dataset timestamps; `REPLAY_RATE` (default x20) compresses wall time only |
| P6 | Kafka: 1 topic `dns.telemetry.v1`, 3 partitions keyed by `client_ip`, ~1 h retention; ClickHouse: `dns_events_raw` (MergeTree by day) + `site_qoe_minute` |
| P7 | Single `docker compose up`: qvac, kafka (KRaft), emulator, agent, wazuh, clickhouse, grafana + idempotent `provision`; env via `.env`; demo profile `--profile demo-video` |
| P11â€“P18 | Vocabulary / 6 PoPs / 4 clientes / zone_mapping.csv / latency profiles / ground_truth / model `QWEN3_1_7B_INST_Q4` / mono-repo `docs/ app/ deploy/ scripts/` / video script |

**Empirical thresholds** (measured from the dataset): longest single label mean 10.5, p90 17, p95 21, p99 36, max 63 â†’ threshold 40 + entropy + repetition is safe. 29.8 % of unique qnames occur once â†’ rarity alone unusable. Top e2ld repeaters: microsoft.com (10 179), google.com (8404), googleapis.com. QTYPE mix: A + HTTPS/TYPE65 dominate.

## Acceptance criteria

- [ ] `docker compose up` brings up the full stack (qvac, kafka, emulator, agent, wazuh, clickhouse, grafana) idempotently.
- [ ] The agent filters the stream with the 5 signals and escalates only rule-satisfying queries to QVAC; benign queries never call QVAC.
- [ ] QVAC verdict JSON propagates; malformed output degrades to `unverified` alerts (ADR-0002).
- [ ] DGA/tunnel/beaconing/typosquat incidents from the attack script appear in Wazuh with correct verdict, confidence, and severity per `local_rules.xml`.
- [ ] Grafana shows per-site QoE (0â€“100, labels) plus latency/NXDOMAIN/saturation culprits for at least one degraded site during E1â€“E5.
- [ ] Evaluation report (from `ground_truth`, emulator-only) with precision/recall/F1, elimination rate, p50/p95 latency, split by zone/pop.
- [ ] Repo accessible + video â‰¤ 5 min in Spanish; README declares all pre-existing bases (qvac, quirk skills, dataset synthetic).

## Risks and open questions

- QVAC inference latency on a laptop may limit the demo replay rate; `REPLAY_RATE` must be tuned so escalation is observed in the video without overloading the LLM.
- Wazuh on limited RAM (4 GiB class devices) may be heavy; a syslog shim path keeps the demo honest without Wazuh resource spikes.
- The route to sub-issues / tickets follows the quirk workflow: this issue holds the context; the two specs below own implementation.

## Implementation and testing decisions

- Seam: the **agent/filter** is the single testing seam â€” unit-test each signal against fabricated event windows; integration-test escalation against the emulator output; accept the QVAC adapter behind a fake for deterministic CI.
- All services run locally via docker-compose; no network egress for inference.

---

## Subtrees (specs)

This epic is decomposed into two specs (sub-issues), each publishing its own tickets as sub-issues:

- **S1 â€” Threat Detection Pipeline** (emulator, dual clock, attack script, Kafka, filter signals, QVAC verdict, evaluation).
- **S2 â€” QoE, Observability and SIEM** (QoE engine 45/35/20, ClickHouse, Grafana dashboard, Wazuh syslog + local_rules, compose/provision).
